### PR TITLE
fix: prevent stuck withdrawals when Election cap exhausts a group

### PR DESCRIPTION
**Pull request by pahor167**
## Summary

Fixes the production bug from [forum.celo.org/t/.../13333](https://forum.celo.org/t/i-encountered-a-bug-while-unstake-stcelo-and-request-assistance/13333) where user `0x85ca0Dff027102ea3FBF1c077524eab21D1F7927` unstaked 146,388 stCELO but only 10,000 CELO became claimable - 152,360 CELO is pinned to validator group `0x81AE1C73A326325216E25ff1af9EA3871195036E` with no recovery path.

### Root cause

`Account.scheduleWithdrawals` validated the per-group withdrawal against `getCeloForGroup`, which counts `scheduledVotes.toVote`. When the Election `numVotesReceivable` cap is fully consumed, that `toVote` can never be activated by the bot, so it inflates the apparent group capacity. Withdrawals pass scheduling but later revert `InsufficientRevokableVotes` inside `Account.withdraw`, with no recourse for the user.

### Fix

- **New `Account.getRealisableCeloForGroup`** = `max(0, [max(0, revokable - toRevoke) + min(Account.balance, toVote)] - toWithdraw)`, where `revokable` is the group's pending+active Election votes. Counts only CELO that `Account.withdraw` can physically deliver right now: revokable Election votes, plus the slice of `toVote` actually backed by `Account.balance` (the immediate-withdrawal path transfers it directly without calling `Election.vote`). Existing `toWithdraw` is subtracted once from the combined capacity (withdraw is immediate-first, so a pending withdrawal costs one unit of total capacity, not one per bucket). Stale `toVote` the Election cap permanently blocks is excluded - it has no balance backing.
- **`Account.scheduleWithdrawals`** now books each per-group entry through the shared `_addBeneficiaryWithdrawal` helper, which validates against realisable capacity and threads a `balanceBudget` across groups. `Account.balance` is global and shared across all groups' `toVote`, so per-group validation alone let a multi-group call double-count the same unlocked CELO; the threaded budget models `Account.withdraw`'s immediate-first consumption exactly. Any pin that schedules is physically fulfillable.
- **`DefaultStrategy.generateWithdrawalVoteDistribution`** and **`SpecificGroupStrategy.generateWithdrawalVoteDistribution`** (overflow branch) cap each group by the realisable view for real withdrawals, so distribution spreads across healthy groups instead of pinning to an over-allocated HEAD. The DS loop skips zero-realisable groups (without consuming an output slot) and skips any group already chosen this call, so it never emits the same group twice (a duplicate would revert downstream `scheduleWithdrawals`). The transfer path (`isTransfer=true`, accounting-only) keeps the stCelo cap.
- **New `Account.rescueScheduledWithdrawal(beneficiary, fromGroup, toGroups, amounts)`** - **permissionless** recovery for an already-stuck pin, callable by anyone IFF `fromGroup` is provably in deficit for this beneficiary: `min(toVote, balance) + revokable < toWithdrawFor[beneficiary]` (`_requireGroupInDeficit`, per-beneficiary not per-group). Bookkeeping only (no CELO moves); net change to `totalScheduledWithdrawals` is zero. Source slot is cleared FIRST so a malformed `toGroups` containing `fromGroup` cannot destroy the beneficiary's claim. Beneficiary identity is preserved end-to-end (CELO can only flow to `beneficiary` via `Account.withdraw`), so funds cannot be stolen or destroyed.

### Upgrade safety

Storage layout preserved across `Account`, `DefaultStrategy`, `SpecificGroupStrategy` (append-only logic + new external/private functions, no new state slots). `DefaultStrategy.generateWithdrawalVoteDistribution` gains a `bool isTransfer` parameter; callers in `Manager` and `SpecificGroupStrategy` are updated. Once deployed, **anyone** can unstick the forum user by calling `rescueScheduledWithdrawal` - no multisig governance vote required.

This PR targets **`releases/4`** (the deployed release line) so it can be released directly. Versions are bumped to exactly what the contract-compatibility check requires against `releases/4`: `Account` 1.2.1.0 → **1.2.2.0** (minor: added `getRealisableCeloForGroup` + `rescueScheduledWithdrawal`), `DefaultStrategy` 1.2.0.0 → **1.3.0.0** (major: `generateWithdrawalVoteDistribution` gained the `isTransfer` arg), `SpecificGroupStrategy` 1.1.1.0 → **1.1.1.1** and `Manager` 1.3.1.0 → **1.3.1.1** (patch). Storage layouts unchanged; no new state slots.

## Test plan

- [x] `test-ts/account-realisable.test.ts` - unit tests for `getRealisableCeloForGroup`, `scheduleWithdrawals` prevention, and `rescueScheduledWithdrawal` (incl. deficit gating and the `fromGroup`-in-`toGroups` edge case).
- [x] `test-ts/end-to-end-stuck-withdrawal.test.ts` - e2e regression with real contracts (no mocks): exact root-cause-shape reproduction (unbacked `toVote` -> `scheduleWithdrawals` reverts `WithdrawalAmountTooHigh`), both distribution paths (`DefaultStrategy` realisable cap + `SpecificGroupStrategy` `GroupNotBalanced`), in-flight-transfer unbacked `toVote`, zero-realisable HEAD skip, no-duplicate-group, rescue gating (reject when liquid `toVote`+balance or revokable covers the claim), and shared balance-budget across multi-group calls.
- [x] Duplicate-group regression (HEAD with realisable far below its stCelo - the skip-already-chosen fix) covered by `end-to-end-stuck-withdrawal.test.ts` "REGRESSION P2 #4". `account-realisable.test.ts` also covers the `toWithdraw`-subtracted-once boundary (`C - W`).
- [x] Full suite: **951 passing** (on the `releases/4` base). `yarn lint:ci` clean. Contract-compatibility check passes against `releases/4`.
- [x] `scripts/verify-stuck-withdrawal-mainnet-fork.sh` - anvil fork of **real Celo mainnet state** (24 assertions, all passing). (A) RECOVERY: fork the historical withdrawal block, replay the user's exact `Manager.withdraw` with the deployed code to recreate the stuck pin (the live pin self-resolved post-incident), then a permissionless `rescueScheduledWithdrawal` from a random EOA moves it off the bad group and a real `Account.withdraw` succeeds. (B1) BUG ENTRY at the user's withdrawal block 67413681: deployed code burns 146,388 stCELO and pins 152,349 CELO to the bad group whose Election-revokable is only 100,853 CELO - permanently unfulfillable, exactly how the user got stuck. (B2) PREVENTION at the same block with all four patched impls: the same `Manager.withdraw` caps each pin at realisable capacity (no unfulfillable pin), and a real `Account.withdraw` succeeds. Requires foundry + a Celo RPC; not part of CI.
- [x] Realisable math: `toWithdraw` subtracted once (not twice) from combined capacity - addresses review feedback; full suite + fork re-verified.
- [ ] Multisig deploys new impl and upgrades the three proxies.
- [ ] Anyone calls `Account.rescueScheduledWithdrawal(0x85ca0Dff..., 0x81AE..., [healthy groups], [amounts])` on mainnet.
- [ ] Bot processes the rescheduled withdrawal across healthy groups; user calls `Manager.claim` after the 3-day unbond.

## Known follow-up (not yet addressed)

- MEDIUM: `DefaultStrategy.generateWithdrawalVoteDistribution` can revert `NotAbleToDistributeVotes` via slot-budget exhaustion (`maxGroupsToWithdrawFrom`) when realisable is fragmented across more groups than the slot limit, even though the funds are physically withdrawable. The realisable cap shrinks per-slot contribution, making this more reachable when realisable is tight.

### CODE REVIEW

File: contracts/Manager.sol
Comment (Mc01) [inline contracts/Manager.sol]: Should version be bumped to
```suggestion
        return (1, 4, 0, 1);
```
?

File: contracts/SpecificGroupStrategy.sol
Comment (Mc01) [inline contracts/SpecificGroupStrategy.sol]: Same question about semver here and in other files

File: contracts/Account.sol [hunk 1]
```diff
@@ -804,6 +882,59 @@ contract Account is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Managed, I
...
+     * more balance-backed capacity than the balance covers; recoverable
+     * via the rescue functions below.
+     * @param group The address of the validator group.
+     * @return The realisable CELO for `group`.
+     */
+    function getRealisableCeloForGroup(address group) public view returns (uint256) {
+        uint256 revokable = getElection().getTotalVotesForGroupByAccount(group, address(this));
+        uint256 toWithdraw = scheduledVotes[group].toWithdraw;
+        uint256 earmarked = scheduledVotes[group].toRevoke + toWithdraw;
+        uint256 revokableAvail = revokable > earmarked ? revokable - earmarked : 0;
```
Comment (Mc01) [inline contracts/Account.sol]: `toWithdraw` is subtracted here...

File: contracts/Account.sol [hunk 2]
```diff
@@ -804,6 +882,59 @@ contract Account is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Managed, I
...
+        // Existing `toWithdraw` also drains `toVote` via the immediate-
+        // balance path. Subtract the already-reserved slice so a new pin
+        // can only claim the remainder.
+        uint256 toVote = scheduledVotes[group].toVote;
+        uint256 availableToVote = toVote > toWithdraw ? toVote - toWithdraw : 0;
```
Comment (Mc01) [inline contracts/Account.sol]: ...and here...

File: contracts/Account.sol [hunk 3]
```diff
@@ -804,6 +882,59 @@ contract Account is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Managed, I
...
+        uint256 balance = address(this).balance;
+        uint256 immediateCap = balance < availableToVote ? balance : availableToVote;
+        return revokableAvail + immediateCap;
```
Comment (Mc01) [inline contracts/Account.sol]: And the sum here can contain in one scenarios:
```
(revokable - (toRevoke + toWithdraw)) + (toVote - toWithdraw)
```
I'm a bit lost there tbh... is subtracting twice by `toWithdraw` expected?

I would believe it could be something like:
```
((revokable - toRevoke) + toVote) - toWithdraw
```
Could you comment a bit on math here?

File: contracts/Account.sol [hunk 4]
```diff
@@ -804,6 +882,59 @@ contract Account is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Managed, I
...
+     * when the bot processes them, so a scheduling-time check must treat
+     * both as spoken-for to avoid over-promising.
+     *
+     * Note: `Account.balance` is global and shared across all groups'
+     * `toVote`. A multi-group withdrawal call could in principle claim
+     * more balance-backed capacity than the balance covers; recoverable
+     * via the rescue functions below.
+     * @param group The address of the validator group.
+     * @return The realisable CELO for `group`.
+     */
```
Comment (Mc01) [inline contracts/Account.sol:858]: Walls of comments like this are a bit too much, especially since they appear in more places even in this file.

Could you try to minimze the comments with AI and make them more readable, concise and direct/clear?

File: contracts/Account.sol (hunk 3 above)
Comment (pahor167) [reply contracts/Account.sol]: Good catch - you're right, subtracting `toWithdraw` twice is wrong. `Account.withdraw` is immediate-first (`immediate = min(balance, toVote)`, then revoke the remainder), so one unit of an existing `toWithdraw` consumes one unit of the **combined** capacity, not one unit of each bucket. The double subtraction under-counted realisable by up to the whole `toWithdraw`, wrongly rejecting serviceable withdrawals on groups that already carry pins (liveness only - it's always conservative, so it never caused a stuck pin).

Fixed in 4825bb4 to subtract once:
```
capacity   = max(0, revokable - toRevoke) + min(balance, toVote)
realisable = max(0, capacity - toWithdraw)
```
Same fix applied to `_addBeneficiaryWithdrawal` (the actual scheduling path); there the shared `balanceBudget` is now charged `min(amount, immediateBucket)` (the most the pin can draw immediate), which is >= the old charge, so strictly safer against cross-group balance double-claims while the capacity check becomes correct. Verified: full suite 982 passing + mainnet-fork (historical bug-entry + prevention + rescue) all pass with the new math.

File: contracts/Account.sol (hunk 4 above)
Comment (pahor167) [reply contracts/Account.sol:858]: Fixed
